Reason waiting (incl. custom reason) UX

weedySeaDragon opened the pull request:
PT Story:  https://www.pivotaltracker.com/story/show/143810729

This is the 2nd PR (the 2nd half) for the story.
(And this is the 2nd version of the 2nd PR for the 2nd half of the story.)

Changes proposed in this pull request:
2. UI: the _reasons for waiting_ has been moved to above the _application status buttons
3.  Uses AJAX (UJS) to update the db immediately when a reason is selected from the list.  The db is also updated immediately when the text for a 'custom/other' reason is entered.

### Screenshots (Optional):

**"Select a reason..." is a placeholder prompt if no reason is set for the application:**

<img width="781" alt="screen shot 2017-06-28 at 1 46 10 am" src="https://user-images.githubusercontent.com/673794/27628429-da98557c-5ba3-11e7-8233-17428b43fd43.png">


<img width="770" alt="screen shot 2017-06-28 at 1 46 20 am" src="https://user-images.githubusercontent.com/673794/27628426-da938934-5ba3-11e7-9bdb-741e283c8ebb.png">

---

**List of reasons is in the correct locale**

<img width="745" alt="screen shot 2017-06-28 at 1 47 36 am" src="https://user-images.githubusercontent.com/673794/27628427-da952334-5ba3-11e7-8aad-10192cd47ec5.png">


<img width="748" alt="screen shot 2017-06-28 at 1 47 28 am" src="https://user-images.githubusercontent.com/673794/27628430-daa186d8-5ba3-11e7-8f1f-23f8fe62b825.png">


---

**The text field for typing in the "other reason" only shows if that reason is selected**

<img width="757" alt="screen shot 2017-06-28 at 1 44 35 am" src="https://user-images.githubusercontent.com/673794/27628312-7a2bca84-5ba3-11e7-9917-a9acd32102fc.png">


<img width="771" alt="screen shot 2017-06-28 at 1 46 39 am" src="https://user-images.githubusercontent.com/673794/27628428-da9568b2-5ba3-11e7-9612-210214e3bdb6.png">

---

#### The reason isn't shown if the status is not 'Waiting for Applicant'
<img width="623" alt="reasonwaiting-hidden-unless-status-en" src="https://user-images.githubusercontent.com/673794/27395497-dc49d9a2-5664-11e7-8ce8-74a0a92002da.png">

---


<img width="845" alt="reasonwaiting-other-reason-text-en" src="https://user-images.githubusercontent.com/673794/27395495-dc4919cc-5664-11e7-9919-3b4443db5264.png">


Ready for review:
@patmbolger @thesuss @RobertCram

weedySeaDragon committed:
add methods for accessing data depending on the locale

weedySeaDragon committed:
add waiting_reason to MembershipApplication

weedySeaDragon committed:
remove methods no longer needed

weedySeaDragon committed:
update locale info: remove 'save reason..' stuff; add placeholder text "select a reason..."

weedySeaDragon committed:
add steps for checking if something is/isn't in a option (select) list

weedySeaDragon committed:
feature: "other/custom reason..." always exists in the list of reasons

weedySeaDragon committed:
feature: admin sets the reason why SHF is waiting or applicant (incl. possibly a custom reason)

weedySeaDragon committed:
add semi-colon to explicitly end a statement

weedySeaDragon committed:
set info needed for 'other reason' and custom; handle XHR (UJS) interaction; DRY create/update error

weedySeaDragon committed:
view, helper: UX for setting the reason for waiting

RobertCram reviewed (dismissed):
Now all (?) the javascript tests timeout for me. I broke off Cucumber after half an hour. Seems the problem shifted from CI to me. Which is a good thing - I know. But still :)

For me, putting the waiting loop in the After('@javascript, @poltergeist') hook makes all tests run (and green) again.

It definitely seems to been an issue where tests are ended while some code is still running. Executing the failing tests on their own makes them pass. And so does waiting after each test for possible running things to stop.

weedySeaDragon commented:
(whew!)

But that just confirms that we should clean up the cucumber steps to make sure we're doing everything we can to avoid those kinds of situations.  (e.g.  the stuff you're working on; using `assert...` statements).

weedySeaDragon committed:
Merge branch 'develop' into sprint17-#143810729-set-custom-reason-UI

weedySeaDragon committed:
Merge branch 'develop' into sprint17-#143810729-set-custom-reason-UI

weedySeaDragon committed:
+ wait for ajax after updating custom reason text

weedySeaDragon committed:
whitespace formatting change just to force another build

want to see if the previous commit -- adding the "wait for ajax requests to complete" really solves the intermittent failing errors; want to see if we the build will fail
